### PR TITLE
Fix Logging for Validation

### PR DESCRIPTION
**Pull request by TonyCongqianWang**
#440 removed the logging of validation. While we almost never use validation it can be useful for debugging edge cases or when using optimizers different than ranger21.

This PR expands the SimpleLineLogger to also log validation metric (i.e. validation loss).

Additionally it uses `dist_sync=False` for `on_step` logging to reduce overhead. To make up for it, it uses on_epoch logging with `dist_sync=True`. This is also more informative for the most people as single batch values can be very noisy.

Lastly I added the lightweight `csv_logger` additionally to the tb_logger. I have never used the tb_logger myself but I think for most a simple csv file is easier to handle

### CODE REVIEW

**Comment by vondele** [comment]
Can you paste the output here for 2-3 epochs?

**Comment by vondele** [comment]
master:
```
run 1:
Epoch  7: 100%|  1526/1526  [00:22<00:00,  67.23it/s, train_loss=0.00528, v_num=33]
run 2:
Epoch  7: 100%|  1526/1526  [00:22<00:00,  67.80it/s, train_loss=0.00542, v_num=36]
```
branch:
```
run 1:
Epoch  7 (Train): 100%|  1526/1526  [00:23<00:00,  62.19it/s, train_loss=0.00550]
run 2:
Epoch  7 (Train): 100%|  1526/1526  [00:23<00:00,  66.75it/s, train_loss=0.00557]
run 3:
Epoch  7 (Train): 100%|  1526/1526  [00:23<00:00,  62.60it/s, train_loss=0.00534]
```

Seems like it is a bit slower?

File: train.py [hunk 1]
```diff
@@ -43,49 +42,138 @@ def on_train_epoch_end(self, trainer, pl_module):
...
-                f"Epoch {trainer.current_epoch:>2}: "
+                f"Epoch {trainer.current_epoch:>2} (Train): "
                 f"{current_step / total_batches:>4.0%}| "
                 f"{current_step:>5}/{total_batches:<5} "
-                f"[{elapsed_str}<{remaining_str}, {rate:>6.2f}it/s, "
-                f"{self.metric_name}={loss_val:.5f}, "
+                f"[{self._format_time(elapsed_total)}<{self._format_time(remaining)}, "
+                f"{rate:>6.2f}it/s, "
+                f"{self.train_metric_step}={loss_val:.5f}]",
                 f"v_num={trainer.logger.version}]",
```
[the remark this answers is not in the document]
Comment (vondele) [reply train.py:112]: that's probably rather a `,` needed instead of a `]`

File: train.py [hunk 2]
```diff
@@ -43,49 +42,138 @@ def on_train_epoch_end(self, trainer, pl_module):
...
+        if trainer.global_rank != 0 or trainer.sanity_checking:
+            return
+
+        pl_module._log_epoch_end(self.train_metric_epoch)
+        train_loss = trainer.callback_metrics.get(self.train_metric_epoch, float('nan'))
+        print(
+            f"Epoch {trainer.current_epoch:>2} (Train): "
+            f"[{self.train_metric_epoch}={train_loss:.5f}]",
+            flush=True
+        )
```
[the remark this answers is not in the document]
Comment (TonyCongqianWang) [reply train.py:130]: While this technically does cause one redundant log call, it is necessary to keep the lightning module work with a logger correctly without the callback and at the same time ensure that the callback prints an up to date value.

There is not really a cleaner solution available because trainer callbacks are always executed before the module cabacks

One extra log call per epoch shouldn't be too bad.

[the remark this answers is not in the document]
Comment (vondele) [reply train.py:130]: Agreed
